### DISCUSSION TITLE
TIM - controlling auxiliary valves (SCV)

### DISCUSSION THREAD

**Discussion by freggala** [Open-Agriculture/AgIsoStack-plus-plus#508] [General]
I've done some experimentation with AgIsoStack++'s TC-client implementation, but now I'm trying to move on to a new challenge:
I aim to control tractor functions (primarily SCVs) from my application, and I've seen the `isobus::ControlFunctionFunctionalities` class and its methods for detecting and configuring individual Functions. However, I don't see how those are meant to be used or, what's more, how to actually control a tractor function, i.e., how to set a particular flow value.
Can you give me a hand on how to get started with this part of AgIsoStack++?
It seems TIM has not yet been completely implemented in AgIsoStack++—am I right, and if so, how significant is the missing piece? Is it possible to gain control over a tractor's auxiliary valves anyway (workarounds)? 

Thank you very much!

**Comment by ad3154** [comment]
Hey there.

The control function functionalities message is part of the required diagnostic messages in ISO 11783-12. It's often displayed for informational purposes I think, and that's about it. It's not meant for configuring anything during run-time operation. The functionalities are, I think defined by an AEF document instead of actually in the ISO standard, which is unfortunate.

TIM is an AEF standard and is not part of ISO 11783. Open-Agriculture is not a member of AEF, so we have no access the standard, or authority to contribute an implementation of it here at least at this time. TIM is a very large set of features with cryptographic authentication pieces that require interaction with AEF to get working, so there's not much we can do about it. If some company out there wanted to sponsor Open-Agriculture to join the AEF, I am not even sure if they'd let us join it to be honest, since we're not a company.

It could be possible to gain access to auxiliary valve controls on _some_ machines via other methods since there are some messages defined in ISO 11783-7 for doing this, but those would be application layer messages, so you'd have to encode them yourself (which is pretty easy) and send them with the stack . Maybe check out [isobus.net](https://www.isobus.net/isobus/pGNAndSPN/?type=PGN) for valve related PGNs.

**Comment by freggala** [comment]
Hey Adrian,
Thank you very much for pointing this out. I've done some research in the meantime, and as it turns out, part 7 is, curiously, the only part of the ISO 11783 standard that I could come by, even though it's only the ancient 2002 version. Supposedly, PGNs 0xFE3X ( Auxiliary valve X command) are what you mean, so I cobbled together a little program (attached), which, as apprehended, puts this command message on the bus but seems to be sadly ignored by my JD 6130R's tractor ECU. 
[main.cpp.txt](https://github.com/user-attachments/files/17885389/main.cpp.txt)
Do you see any obvious flaws in my approach or have any ideas about what else to try?

I'm also looking for ways to circumvent the problem, i.e., to control the SCVs indirectly instead. Are AUX functions a viable option here? (I may have misunderstood the concept of AUX functions....)
